### PR TITLE
Harden

### PR DESCRIPTION
**Pull request by NeQuissimus**
- [X] Switch all URLs to use HTTPS
- [X] Verify SBT launcher using signature and MD5 hash
- [X] Add -skip-verify to skip artifact verification
- [X] Added -skip-verify to download tests, I was unable to stub out md5sum, this should probably be fixed in a follow-up

The three GPG keys listed are the only ones in the entirety of http://repo1.maven.org/maven2/org/scala-sbt/sbt-launch/

I do not have a macOS machine handy, so I do not know if this works there, needs testing.

/cc @dwijnand @alexarchambault

### CODE REVIEW

File: sbt [hunk 1]
```diff
@@ -130,10 +133,10 @@ make_url () {
...
   case "$version" in
         0.7.*) echo "$base/files/sbt-launch-0.7.7.jar" ;;
-      0.10.* ) echo "$base/org.scala-tools.sbt/sbt-launch/$version/sbt-launch.jar" ;;
-    0.11.[12]) echo "$base/org.scala-tools.sbt/sbt-launch/$version/sbt-launch.jar" ;;
-          0.*) echo "$base/org.scala-sbt/sbt-launch/$version/sbt-launch.jar" ;;
-            *) echo "$base/org/scala-sbt/sbt-launch/$version/sbt-launch.jar" ;;
+      0.10.* ) echo "$base/org.scala-tools.sbt/sbt-launch/$version/sbt-launch-${version}.jar" ;;
+    0.11.[12]) echo "$base/org.scala-tools.sbt/sbt-launch/$version/sbt-launch-${version}.jar" ;;
+          0.*) echo "$base/org.scala-sbt/sbt-launch/$version/sbt-launch-${version}.jar" ;;
+            *) echo "$base/org/scala-sbt/sbt-launch/$version/sbt-launch-${version}.jar" ;;
```
Comment (dwijnand) [inline sbt]: This change will break the <1.0.0 jars.

What is the motivation for this change?

File: sbt [hunk 2]
```diff
@@ -278,9 +281,36 @@ acquire_sbt_jar () {
   } || {
     sbt_jar="$(jar_file "$sbt_version")"
     download_url "$(make_url "$sbt_version")" "$sbt_jar"
+    [[ -n "${skipVerify}" ]] || verify_sbt_jar "$sbt_jar"
```
Comment (dwijnand) [inline sbt]: Let's do things incrementally.  Let's skip gpg key retrieving, verifying and skipping for now and just do the HTTPS change and the MD5 checking.

File: test/test_helper.bash [hunk 3]
```diff
@@ -1,7 +1,7 @@
 export TEST_ROOT="$BATS_TMPDIR/$(basename "$BATS_TEST_FILENAME").$$"
 export TEST_BIN="$TEST_ROOT/bin" && mkdir -p "$TEST_BIN"
 export HOME="$TEST_ROOT" # todo - eliminate
-export PATH="$BATS_TEST_DIRNAME/../bin:$TEST_BIN:/usr/bin:/usr/sbin:/bin:/sbin"
+export PATH="$BATS_TEST_DIRNAME/../bin:$TEST_BIN:/usr/bin:/usr/sbin:/bin:/sbin:${PATH}"
```
[the remark this answers is not in the document]
Comment (NeQuissimus) [reply test/test_helper.bash]: I can take this out and just add it back manually whenever I need it... I am fine with that.

File: sbt (hunk 1 above)
Comment (NeQuissimus) [reply sbt]: The JARs without version number do not have signatures.
The ones with version number are not identical to the ones without (I assume they are recompiled from the same code), so we cannot use the versioned signature to verify the unversioned JAR.

File: test/download.bats [hunk 4]
```diff
@@ -34,14 +42,15 @@ no_properties_and_fetch ()               { assert_no_properties && fetch_launche
 fetch_launcher () {
   local version="$1" && shift
   stub_curl
+  stub_curl
```
Comment (NeQuissimus) [inline test/download.bats]: @dwijnand Something in this test framework is strange. I can't figure it out... It seems I have to stub curl twice, once for each potential invocation (curl JAR + curl MD5).
But I cannot keep the unstub instances. Those will fail if I keep them around.

Similarly, the unstub in the `teardown` throws an error if you get rid of the `rm` line.

File: sbt (hunk 1 above)
Comment (dwijnand) [reply sbt]: I see!

Let's push this change to later when we consider adding signature verification.

```suggestion
            *) echo "$base/org/scala-sbt/sbt-launch/$version/sbt-launch.jar" ;;
```

File: test/download.bats (hunk 4 above)
Comment (dwijnand) [reply test/download.bats]: Yeah, it's confusing as hell. 😄

What do you mean by "I cannot keep the unstub instances"?

I can have a look at this detail (if you want) after you get back to me.

File: sbt [hunk 5]
```diff
@@ -255,9 +255,11 @@ download_url () {
   local url="$1"
   local jar="$2"
 
-  echoerr "Downloading sbt launcher for $sbt_version:"
-  echoerr "  From  $url"
-  echoerr "    To  $jar"
+  if [ -z "$3" ]; then
```
Comment (dwijnand) [inline sbt]: Rather than this, let's split the message from the curl/wget utility.

File: sbt [hunk 6]
```diff
@@ -278,9 +280,28 @@ acquire_sbt_jar () {
...
+  local jar="${1}"
+  local md5="${jar}.md5"
+
+  download_url "$(make_url "${sbt_version}").md5" "${md5}" true > /dev/null 2>&1
+  if echo "$(cat "${md5}")  ${jar}" | md5sum -c -; then
+    rm -f "${md5}"
+    return 0
+  else
+    echoerr "Checksum does not match"
+    return 1
```
Comment (dwijnand) [inline sbt]: If I follow the logic properly this will result in "Checksum does not match" and then "Download failed. Obtain the jar manually and place it at $sbt_jar".  If checksums don't match is it fair to say "Download fail"?  I'm genuinely not sure.  WDYT?

File: sbt (hunk 1 above)
Comment (NeQuissimus) [reply sbt]: You can't the same is true for the files with the md5 hashes

File: test/download.bats (hunk 4 above)
Comment (NeQuissimus) [reply test/download.bats]: I removed all the `unstub curl` stuff because it will do an `exit 1`, which fails the test.
I could not figure out how to make "unstubbing" work.

File: sbt (hunk 6 above)
Comment (NeQuissimus) [reply sbt]: Maybe re-phrasing the message may be better.
We could not download *and validate* the file.

File: sbt (hunk 1 above)
Comment (dwijnand) [reply sbt]: Hmm.  Well, that's concerning.

If you don't mind, I'm going to keep this subthread open as I want to understand why or, more importantly, how they're different.

File: sbt (hunk 6 above)
Comment (dwijnand) [reply sbt]: Yeah, sounds good!

File: sbt (hunk 1 above)
Comment (dwijnand) [reply sbt]: I actually jumped on this straight away.  Using https://github.com/scala/jardiff there's no difference.  From vimdiff it looks like it's just timestamps.  So I guess we're fine to make this change now.

File: test/download.bats (hunk 4 above)
Comment (dwijnand) [reply test/download.bats]: OK, I'll have a look.

File: sbt [hunk 7]
```diff
@@ -277,10 +273,35 @@ acquire_sbt_jar () {
...
+    esac
   }
 }
 
+verify_sbt_jar() {
+  local jar="${1}"
+  local md5="${jar}.md5"
+
+  download_url "$(make_url "${sbt_version}").md5" "${md5}" > /dev/null 2>&1
+  if echo "$(cat "${md5}")  ${jar}" | md5sum -c -; then
```
Comment (dwijnand) [inline sbt]: General review notes / TODOs / "let's not forget"s

* What about md5sum's stdout/stderr output?
* What about if `md5sum` isn't available, but `md5` is?
* Or if none are available?

File: test/download.bats (hunk 4 above)
Comment (dwijnand) [reply test/download.bats]: Hey, @NeQuissimus, sorry for the delay.

So the `stub` machinery is very strict.  Once you use a stub, it's used, which is why you need two stubs for each curl usage _and_ `unstub` is important as it verifies that you used everything that you stubbed.

I've got a local version working where the fundamental diff is this:

```diff
-stub_curl() { stub curl "$curl_opts"; }
+stub_curl() {
+  stub curl "$curl_opts";
+  case "$1" in
+    0.*) ;; # sbt <1 has no MD5 files to curl and verify
+      *) stub curl "$curl_opts" && stub md5sum 'true' ;;
+  esac
+}
```

Let me know how that works, I'll circle back to this in a few days.

File: sbt (hunk 7 above)
Comment (NeQuissimus) [reply sbt]: `md5sum` is in `coreutils`, are those ever not available? macOS?

Comment (dwijnand) [reply sbt]: So it looks like at the very least it's `md5sum` on both macOS and Linux (as opposed to `sha1sum` on macOS vs whatever it is for Linux).

But the only way to have the `md5sum` on macOS is by installing it (such as `brew install coreutils`).  We should decide and implement how sbt-extras should behave when it's not present.

Comment (NeQuissimus) [reply sbt]: Is there a built-in equivalent on macOS?

Comment (dwijnand) [reply sbt]: Oh!  It seems there's BSD's `md5` in `/sbin/md5`.  Probably should fall back to that.  And then still decide what to do if neither is present (e.g. on Cgywin).

Comment (NeQuissimus) [reply sbt]: I'll do that. I imagine not having such a command means failure... Same as if the hash didn't match. After all, we could not validate the hash...

**Comment by NeQuissimus** [comment]
Alright, how about this?
All the `unstub` are back and I added the check for the existence of an MD5 command.

File: sbt [hunk 8]
```diff
@@ -277,10 +273,44 @@ acquire_sbt_jar () {
...
+    md5_cmd="md5sum"
+  elif command -v md5 > /dev/null 2>&1; then
+    md5_cmd="md5"
+  else
+    echoerr "Could not find an MD5 command"
+    return 1
+  fi
+
+  download_url "$(make_url "${sbt_version}").md5" "${md5}" > /dev/null 2>&1
+  if echo "$(cat "${md5}")  ${jar}" | "${md5_cmd}" -c -; then
```
Comment (dwijnand) [inline sbt]: This PR is looking really good now, thank you!

Two things:

First, how does a user without md5 _or_ md5sum progress?  With the current patch what they can do is download the jar and feed it to `-sbt-jar`.  Is that good enough?  Is it silly to have a `-sbt-jar-md5`?  🤔 let me know if you have any thoughts on this.

The second one is `md5sum -c -` works but `md5 -c -` isn't.  Here's what I've found works with `md5`:

```bash
09:22:42 $ if [ "$(md5 -q sbt-launch-1.2.8.jar)" == "$(cat sbt-launch-1.2.8.jar.md5)" ]; then echo rm; else echo fail; fi
rm
```
